### ISSUE TITLE
Talk about features that would help Single Page Apps

### ISSUE THREAD

**Issue by jeffcarp** [braintree/braintree-web#29]
`braintree-web`'s design comes from a time when checkout flows generally used a `<form>` POST. This is no longer the case in 2015, where SPAs and JS-heavy checkout flows are common.

There are a couple things that keep `braintree-web` from working well in SPAs, the biggest of which is that `braintree.setup` can only be called once per page load.

I'm starting this thread to discuss what kinds of features would be useful for making `braintree-web` work better in JS-heavy checkout flows.

Related Issues:

- setup() with different user #18
- Ability to disable dropin form event handlers #4
- paymentMethodNonceReceived callback doesn't get a new nonce after submitting and re-entering number #20
- paypal support? https://github.com/jeffcarp/braintree-angular/issues/8

**Comment by trainerbill** [comment]
I have actually had a lot of success using Angular and braintree-web.  I didn't know that braintree-angular exists so I will need to take a look at that.  A couple things I would recommend are:

1.  paymentMethodNonceReceived: function (event, nonce).  It would be nice if this function returned a creditcard object containing like last 4 of card etc so that we do not need to vault the customer/method to display it on a confirmation route.  Obviously no full informaiton for PCI.  Similar to how the paypal button returns shipping and email info 

2.  Obviously the event handlers on forms have been a problem.  I have my braintree.setup calls in a service and it seems like I can use JQuery to empty() the drop in container and call it again.  Would be nice to see this reworked so that we can call setup multiple times without any work arounds.

3.  Need a bower install for braintree-data.  Also need a callback function on in the BraintreeData.setup call so that it returns the data instead of creating a DOM element in the form.

There are more that I ran into so I will update as I run into them.

**Comment by kyledetella** [comment]
@trainerbill regarding `paymentMethodNonceReceived`, as of `v2.5.0` we added a new root level callback called `onPaymentMethodReceived` which returns more details for Credit Card transactions. Unfortunately, we have yet to release the official documentation, but you can read more about the callback [here](https://github.com/braintree/braintree-web/issues/15#issuecomment-74783073) and I am happy to answer more any further questions you may have.

**Comment by trainerbill** [comment]
Awesome!  Thanks for the heads up on that.

**Comment by tnunes** [comment]
:+1: for a quick resolution of this. These issues make it really difficult to integrate braintree in a SPA.

**Comment by filipediasferreira** [comment]
We appreciate a quick fix to ease integration.

**Comment by joanamelo** [comment]
:+1:

**Comment by carsy** [comment]
:+1:

**Comment by tnunes** [comment]
We were experiencing an issue where loading the braintree widget (with `braintree.setup()`) for a second time on the same single page app (after removing the previous one, but without reloading the page), would make the widget use the client token with which `braintree.setup()` was called for the first time, effectively giving access to the vault of the first user.

This is a serious issue, since it could allow a user A access to the vault of a user B after user A logged out, if the page was not refreshed meanwhile.

After investigating the problem more thoroughly, we were able to work around the issue, but we had to resort to some nasty hacks and fragile solutions that might break with future releases of `braintree-web`.

We found that all the following steps are required for a proper cleanup and new setup of the widget:

- remove the DOM element where the widget is attached;
- remove the `"braintree-dropin-modal-frame”` iframe added to the body;
- delete `window.Braintree` object;
- delete various `window.callback_json_xxxx` functions (2 per `setup()`);
- remove various “on message” event listeners attached to the `window` object (5 per `setup()`);
- remove the script from the page (since we use `require.js`, we’re resorting to `requirejs.undef('braintree’)` to do this).

After all these steps are done, we can then require the `braintree.js` script again and invoke `braintree.setup()` with different parameters and user vaults and payments work as expected.

To remove the event listeners left hanging in the `window` object we resorted to monkey-patching the `window.addEventListener` method, which feels wrong and may lead to other issues.

A `destroy()` method implemented by `braintree-web` has the benefits that it doesn’t need to resort to nasty hacks and monkey-patches to keep track of callbacks and event handlers that are registered globally, since the module can keep track of its own references for easier removal.

If you could provide a `braintree.destroy()` (or similar) method that would do the steps described, integration in single page applications would be much easier and we wouldn’t have to fear that future upgrades of `braintree-web` could break our workaround solution.

Reloading of `braintree.js` should be left to consumers, since the method to do that depends on how the script is included, but all the other steps could be provided by your own cleanup mechanism.

**Comment by imgntn** [comment]
wow @tnunes i was trying the same kind of workaround a couple of days ago, but missed the callbacks and event listeners on the window.  thanks for the info.  do you think the hosted fields would have the same issues?

**Comment by tnunes** [comment]
Hi @imgntn,

I haven't used hosted fields, so I can't help you that. I know, however, that `braintree-data` also presents the same kind of issues, since to integrate it I also had to find out and clean what it leaves attached to the page.

**Comment by ms7s** [comment]
Since github doesn't support voting for issues, I'm adding a comment: Please add `braintree.destroy()`.

**Comment by karpitsky** [comment]
+1

**Comment by niklr** [comment]
We NEED braintree.destroy()!

**Comment by fanfreeman** [comment]
Really need this. Stripe has it, but we don't want to give up on Braintree just because of this issue.

**Comment by baillyje** [comment]
+1

@tnunes, Do you know if it's possible to target the braintree 'on message' events listener without deleting all others ? Any Ideas of their name ?

**Comment by tnunes** [comment]
What I do is monkey-patch the `window.addEventListener` function right before I setup braintree.

In my implementation I accumulate all event listeners in an array before proxying to the real `window.addEventListener`.

Then, in my `braintree.destroy`, I use `window.removeEventListener` to remove all listeners I previously saved references to, right before I restore the original `window.addEventListener` function.

This is hackish and fragile, but as long as nothing aside from braintree adds listeners to the page from the moment you call `setup` until you destroy the widget, you should good.

**Comment by HiroAgustin** [comment]
We have the same problem @tnunes described, thank you for the hack. Still, I think we are going to do an full page reload until `braintree.destroy` (or something similar) is implemented by default.

**Comment by tyleriguchi** [comment]
Glad I stumbled upon this issue, was tearing my hair out trying to figure out why I was getting errors in my Ember app when I tried to add and remove paypal accounts. A teardown method would really help out for single page apps.

**Comment by schmitzdenis** [comment]
+1

**Comment by paulrnash** [comment]
+1 come on braintree, we're all waiting!

**Comment by ThoamsClark** [comment]
+1

Would be really helpful. Reloading the whole page seriously detracts from our whole checkout experience.

**Comment by tastyeggs** [comment]
+1

Come on braintree, get your act together. Come live in the future with us. Everyone is designing single-page applications and it is extremely frustrating when we have to integrate with libraries designed by companies who think they use the latest tech, but actually have no clue.

This is CS 101 fundamentals; make your code self-contained; make it leave no trace on the system its running on. We thought we should use braintree because it would be easier to integrate, but it's taking the same amount to time, and same amount of nasty hacks as it takes to integrate with your parent company.

You sitting on this issue for 3 months shows that you either don't care about your users experience, or you have no clue what single-page applications are about, hence don't see how badly you're crippling the experience of our users and developers. Either way, it reflects poorly on your company.

Forgive my strong language; but this is *exactly* how I feel. If this were an open-source library, I would have fixed this by now. But now we're all trapped to writing shitty code, because either you guys don't care, or you guys just don't get it.

Wake up!!

**Comment by paulrnash** [comment]
Pretty much what he said. I've wasted hours and hours trying to make
something reasonable out of this technology. It doesn't seem like either
the default web integration or hosted fields will work for anything but a
multiple page click through experience. After earlier experiences where
braintree's support had been impeccably good, I am now finding that it is
not so responsive and technically they are not keeping up with the times.
Perhaps all the PayPal and eBay split noise is affecting the focus of the
team or confusing the priorities. I am seriously considering switching to a
competitor after my experience dealing with this mess, particularly sense
the communication from Braintree has mostly been either radio silence or
"uh sorry." they don't even seem to be aware of the reason this is such an
inconvenience.

**Comment by rturumella** [comment]
Hey folks - I'm a Product Manager with the Braintree SDK and Developer Experience team. 

We understand your frustration and the importance of providing an experience that works in a myriad of integration scenarios, including Single Page Applications.

We also recognize that this is an area in which we can improve our SDK. We're currently actively exploring the features we need to add and making considerations to preserve compatibility for all braintree.js users. 

This is our best path forward for us to provide a more ideal and complete experience for all of our merchants. I'll keep you posted on the progress of this as I know that it is currently resulting in a less than ideal integration experience.

In the meantime, take a look at the docs for our [advanced integration](https://developers.braintreepayments.com/javascript+node/guides/credit-cards#tokenize-card) or feel free to contact me (rohit@getbraintree.com) if you have further questions or concerns.

**Comment by tastyeggs** [comment]
@rturumella There's no information in your documentation on how we can do custom integration with PayPal, other than this one line here:

https://developers.braintreepayments.com/javascript+dotnet/guides/paypal/overview

"Braintree offers a few options when accepting PayPal payments. First, you'll need to choose whether you would like to use our Drop-in UI or a custom integration. Learn more about differences between these integrations".

**Comment by tastyeggs** [comment]
@rturumella Any chance you guys can publish the un-minified source of the JS client? Might just make it easier for us to workaround this issue.

**Comment by rturumella** [comment]
@tastyeggs For each release of braintree.js we also have an accompanying unminified file. The pattern is as follows: `https://js.braintreepgateway.com/js/braintree-<VERSION>.js`. 

That file is built from CommonJS modules and as such is not well suited for an open source release. I would recommend taking a look through the latest JS file and perhaps posting your specific issue here or emailing us to see if we can narrow down the exact problem you are facing with your integration.

I would also take a look at the following link:
`https://developers.braintreepayments.com/javascript/guides/paypal/client-side`

**Comment by davidbryant** [comment]
+1

Since the drop-ui doesn't work for single page apps, I'm trying to integrate using the advanced integration to tokenize the cards, but finding the documentation somewhat lacking in detail. Do you have some full docs for your JS API anywhere? 

Right now it's the tokenizeCard() method that I'd like some more details on:

* Can you provide more than just the postal code for address verification? If so how? Is it the same format as the server side deprecated CreditCard.create() function?
* What validation is token by the tokenizeCard() method? Is it verifying the card number when it is called, or is that delayed until a payment is made?
* Can you provide some documentation on the expected error formats returned from that API? None of the minimal examples for this show any error handling at all.

**Comment by paulrnash** [comment]
For your first question, the link
<https://developers.braintreepayments.com/javascript+node/guides/credit-cards#tokenize-card>Rohit
gave seems to have what you want, it shows several parameters to
tokenizeCard such as cvv, etc. As for when/if the validation happens, it
depends on a number of things. There is an option called verify_card that I
believe you can pass in there. It also depends on whether you have
configured the global verification settings for your account.

Braintree do document the error objects in their online docs, but you're
right that they don't show much in the way of error handling best
practices. :(

Now the bigger problem -- what Rohit hasn't mentioned here is that it's
unclear how much a custom integration actually helps us. With a custom
integration, customers are no longer able to qualify for SAQ-A PCI
compliance, which is why they are working on their "Hosted Fields" product.
The problem is that Hosted Fields has all the same problems with Single
Page Applications that the dropin integration has (and then some). Now way
to clear or reset it, or to accept a second payment entry after the first
one has been completed.

**Comment by stubar** [comment]
+1 for this, I went down the route of manually trying to teardown braintree with limited success, seemed top fragile to risk it, so now using an angular directive I detect if braintree as already being initialised if so I add a target='_self' to any add card adding link thus reloading the page.  there's also a check in my card service  so if someone hits the back button it will reload the page.

**Comment by cityreader** [comment]
@tnunes I am trying to set up braintree hosted fields more then once in a Drupal commerce type. So when user changes payment method hosted fields will be loaded via Ajax. The first time load always works, but not the second time. I am trying to reset Braintree by following your instruction above. But there is no avail. I cannot removeEventListener at all and I found this may be related to framebus. Could you show me the code to monkey patch `window.addEventListener` please? Much appreciated:D

**Comment by duizendnegen** [comment]
+1

currently simply recreating the vault and manually filtering out already used nonces (but that doesn't circumvent a 2nd user having access to the 1st users vault)

**Comment by tnunes** [comment]
@cityreader I'm not using hosted fields, so I do not know if they need any additional house keeping.

To monkey patch `window.addEventListener` you can use something like:

```js
var realWindowAddEventListener = window.addEventListener,
    braintreeEventListeners = [];

window.addEventListener = function () {
    braintreeEventListeners.push(Array.prototype.slice.apply(arguments));
    realWindowAddEventListener.apply(window, arguments);
};

// setup braintree

// then to cleanup:
// - do the other described tasks;
// - remove listeners and restore `window.addEventListener`:
var listenerArgs;

while (var listenerArgs = braintreeEventListeners.pop()) {
    window.removeEventListener.apply(window, listenerArgs);
}

window.addEventListener = realWindowAddEventListener;
```

**Comment by paulrnash** [comment]
I gave up on hosted fields, fwiw. I've resigned myself to have to do the
more elaborate SAQ for a custom integration (which will require scanning).

**Comment by cityreader** [comment]
Thanks @tnunes. This approved that I did right before according to your instructions. Sadly, that does not work for `Hosted Fields`.

It seems `Hosted Fields` script binds events in other places including `form` and submit buttons other than `window`. I cannot make multiple setup work unless hacking the braintree hosted fields script, but even in that way event still cannot be unbinded.

So in the last, we decided to reload page when user change a different payment method by clicking radio options. In order to achieve that, we still have to unbind form submit handler binded by Braintree.

This is no way to remove anonymous callback function, so we have capture all listener as well. Instead of monkey patching `window.addEventListener`, we make try to monkey patching `Element.prototype.addEventListener`:

~~~~
var braintreeListener = function() {
  this.originalAddEventListener = Element.prototype.addEventListener;
  this.listeners = [];
  var self = this;

  Element.prototype.addEventListener = function (event, listener) {
    arguments = Array.prototype.slice.apply(arguments);
    // Apply original attach listener function 
    self.originalAddEventListener.apply(this, arguments);
    // Include element data 
    arguments.unshift(this)
    self.listeners.push(arguments);
  };

  return this;
}();

braintreeListener.undelegate = function (html) {
  arguments = Array.prototype.slice.apply(arguments);

  if (typeof html == 'string') {
    html = document.querySelector(html);
  }

  for (var key in this.listeners) {
    if (html == this.listeners[key][0]) {
      // Remove handler
      html.removeEventListener(this.listeners[key][1], this.listeners[key][2]);
    }
  }
}
~~~~

So run this script right before Braintree Hosted Fields.js, when we want to unbind form submit handler we run 

~~~~
$form = $('#form-id');
braintreeListener.undelegate($form[0]);
~~~~

**Comment by iokaravas** [comment]
I ended up "solving" this by using an iframe. It might not seem clean but it works perfectly fine and there are no left-overs.

I'm creating a new iframe, injecting html that includes the form and script setup for braintree with the Client Token using ``` javascript:window["content"] ``` .

Then every callback of braintree's drop-in (onPaymentReceived, onReady, onError) is handled by using ``` Window.postMessage() ``` on the main application window.

Every time you need to create a new drop-in with a different Client Token, you just remove the iframe, and re-create it. This is untested as far as older browsers go but it works perfectly fine on Firefox and Chrome.

Let me know what you think.

**Comment by fadomire** [comment]
Same problem here missing a braintree.destroy method to remove it cleanly and reuse it later without page reload

**Comment by mbrochh** [comment]
This is extremely frustrating. Please, Braintree, give us a `destroy()` method. I have already wasted more than 10 hours on this and can't find a clean solution for hybrid mobile apps...

**Comment by stubar** [comment]
This is interesting, I ran this idea past Braintree's technical support a
while back and they said to me 'others have tried similar work-arounds to
little success', so I was put off the idea.  I wish you luck and hope to
hear you prove them wrong because it sounds like the most reliable method
I've heard so far.

**Comment by gambry** [comment]
+1 for destroy(), IMHO the most important feature to be implemented so far.

**Comment by marcucio** [comment]
+1 for braintree.destroy()!

**Comment by paulrnash** [comment]
Hey @rturumella,

6 weeks or so ago you said you'd keep us posted. Do you have anything you
can share about plans to address our concerns? Like a specific date? It
strikes me that it is likely 6 *days* of work on your part to make
improvements, but there is no sign of any changes... Is it important?

**Comment by poscar** [comment]
+1 .destroy() is necessary.

**Comment by ppcano** [comment]
@rturumella @jeffcarp  is there any plan to include `destroy` or an official workaround?

**Comment by Streemo** [comment]
@tnunes  Funny, I was searching for a destroy() method and I couldn't find it.

I have the exact same issue. We need destroy().

This is a huge security flaw.

**Comment by rturumella** [comment]
Hey folks,

I'm a Product Manager with the Braintree SDK and Developer Experience Team. 

We're currently still actively working on this. There are a number of things to keep in mind in order to ensure a first class experience for all braintree.js users. 

We understand the frustration around this and we know this is an area of improvement for our SDK. Please stay tuned for more updates.

**Comment by paulrnash** [comment]
Do you have a schedule?

**Comment by berendberendsen** [comment]
@rturumella that looks like the same response as a while ago, is there a schedule for this?

**Comment by theotherzach** [comment]
@rturumella I'm a frontend developer for Splice and we have a single page Angular app. Should we be switching away from Braintree now?

**Comment by paulrnash** [comment]
It seems so... Braintree seem unable to implement this functionality in any
reasonable timeframe, and equally incapable of even providing an estimate.
It's pretty sad.

**Comment by ThoamsClark** [comment]
Have put live their Hosted Fields integration - while still in beta. Sure there are issues, but for what it's worth the company was just bought by PayPal. They're probably doing something right. 

Their support team is both enthusiastic and helpful, both on the phone and via email. They've acknowledged the need for the above requested functionality, and apologies but 'reasonable timeframe' is utterly subjective and apparently ignorant of the realities of software development lifecycles. Not trying to have a dig, but patience is a virtue.

**Comment by paulrnash** [comment]
Thomas, thanks for the "not a dig". You're apparently unfamiliar with "
planning." It's completely sane and within the norms of professional
software development to ask for a timeframe other than "soon." Especially
when we've been told "soon" several times. Any developer who has ever
gotten paid for what they do would understand why getting information on
schedules is important.

You're right, their email support in general is very helpful. For things
that they have an answer to. Their support has been completely unhelpful,
both via web and via email on this topic. How patient do you think we
should be? If you look, you'll see that this discussion is many months old
already.

BTW, Paypal bought Braintree very nearly 2 years ago, and in any case
getting acquired is neither an excuse for not responding to your customers
nor an endorsement of "doing something right."

If you read the posters here, I think you'd get the sense that we all
*like* Braintree, which is why it's so frustrating that their stuff doesn't
work well, and they aren't responsive about it. You don't need to apologize
for them or question the merits of our complaints. Unless perhaps you work
there - I hope not.

**Comment by theotherzach** [comment]
Apologies, I should have been more detailed in my question. Is there a
means of using Braintree in a single page app that doesn't require any of
the following:

* a full page refresh
* an iFrame
* monkey patching

Because we're currently evaluating Braintree and if the only means of using
their software is one of the three above approaches, then they literally do
not support single page apps.

**Comment by ThoamsClark** [comment]
Hi Theotherzach, thanks for your insight. I don't work for Braintree, and wouldn't have posted something so proselytising if I did. I take your points and admit that I was unaware that the takeover was quite so long ago.

Let's not and say we did, eh?

**Comment by ThoamsClark** [comment]
There isn't a way which I'm aware of which meets your requirements, and I've settled - for now - with refreshing. It isn't perfect, but our business requirements are met.

**Comment by paulrnash** [comment]
Fair enough - we can all agree it's a shame we're even having this
conversation, I'm sure.

FWIW, I eventually gave up on this approach after talking with Braintree's
PCI compliance company, who by the way were very responsive and helpful
too. Apparently the difference between the two levels of SAQ is not
actually that huge of a deal, so I just stuck with my custom integration
and decided I'll have to get my website scanned (by them, for free).

**Comment by wmadden** [comment]
This is the worst

**Comment by trainerbill** [comment]
This is kind of ridiculous.  @tnunes pretty much laid out what needs to be done.  If they published the unobfuscated code, I am sure the community could knock this out in a day.

**Comment by theotherzach** [comment]
I entirely understand their desire to not open source code as a payment provider. However, it would be nice if they would put a warning in their docs that the hosted solutions are not designed to work in single page apps. Because the way I found out about it was when our QA team was reporting double purchases, thank goodness our QA team rocks and that didn't get out into production. I've got a bunch of guards in place now to prevent disaster, but I was expecting to come out here and read about how I was doing it wrong.

I was really hoping to come out here and read about how I was doing it wrong...

**Comment by duizendnegen** [comment]
won't it be possible to create a custom integration in JS together which is designed for single page apps, hooks in just like drop-in?

**Comment by rturumella** [comment]
Hi folks - I'm a Product Manager at Braintree. 

The `.destroy()` (or similarly named API) functionality is coming next week. We’ll be pushing it as an edge branch so you all can at least kick the tires and see progress. Thanks for your patience.



> Braintree seem unable to implement this functionality in any reasonable timeframe, and equally incapable of even providing an estimate. It's pretty sad.

Yes. We should have precisely communicated our expected timing, and we should have been transparent about what was going on here internally while you all have been waiting. 

This change has been more challenging than we anticipated. Since this library supports a variety of integrations across many merchants, the change must not break currently-integrated merchants. Maintaining backward compatibility involved some major internal refactoring, particularly around messaging and iframe orchestration.

That's the explanation for the delay, but it's not an excuse. We did a poor job communicating about the work and the delays, and as a result we let you all down. For that, we apologize. 

We're taking this thread and your feedback seriously. Once we've pushed `.destroy()`, we will hold an internal post-mortem on this issue so we can (a) understand what went wrong in both the delay and our communication about the change, and (b) improve our processes and transparency going forward.

Thank you for your honesty and directness on this issue, and for working with us as we improve the JS library to support the growing number of ways people want to accept payments.

**Comment by theotherzach** [comment]
@rturumella Rad, thanks!

For me, the biggest issue was not knowing there was a single page app issue when adding the drop-in to our page. I was lucky enough to see a `nonce can only be used once` message which tipped me off to the problem, but QA still caught other instances where our users could make double purchases. 

I would really like to see `.create` throw AT LEAST a `console.warn` if it's called more than once, and ideally throw an error. I would also like to see your drop-in docs reflect the actual state of single page apps so users like me don't stumble onto the issue.

**Comment by burnto** [comment]
@theotherzach those are both great suggestions. Thanks for this feedback.

**Comment by shermandickman** [comment]
@rturumella We're dead in the water until .destroy() is released. Any news to share at this time?

**Comment by paulrnash** [comment]
They told us Friday I think that it'd be coming this week... not sure if you saw that already.

-P

**Comment by rturumella** [comment]
@shermandickman The .destroy() (or similarly named API) functionality is coming next week. We’ll be pushing it as an edge branch so you all can at least kick the tires and see progress. Thanks for your patience.

**Comment by wmadden** [comment]
@rturumella thanks very much for the explanation.

FWIW the worst part of this for us - the scariest part - was when we
realized that because that nonce was still present even after the user had
logged out, any other person could have used our app to make purchases with
the previous user's payment method if they got ahold of the device.

**Comment by kyledetella** [comment]
As @rturumella mentioned last week, we are delivering a beta build of `braintree.js` that exposes the ability to tear down an integration and re-create a new one during a lifecycle of a single page view.

Right now this new version is deployed as a `beta` build and can be downloaded from here:

- [https://js.braintreegateway.com/js/braintree-2.13.0-beta.js](https://js.braintreegateway.com/js/braintree-2.13.0-beta.js)
- [https://js.braintreegateway.com/js/braintree-2.13.0-beta.min.js](https://js.braintreegateway.com/js/braintree-2.13.0-beta.min.js)

We will be adding full documentation in an accompanying production release in the coming weeks. In the meantime, we would love to take in some feedback on this new functionality. We know how important this is and we want to make sure when it goes out in production that the experience is rock solid.

Since version [2.7.0](https://github.com/braintree/braintree-web/blob/master/CHANGELOG.md#270), the configuration object passed during `braintree.setup` has provided an `onReady` callback. As of this [latest build](https://js.braintreegateway.com/js/braintree-2.13.0-beta.min.js) the `onReady` callback will provide an integration object on which you can perform actions such as `teardown`. A basic example of this looks like:
	
```js
var myIntegration;
	
braintree.setup(TOKEN, 'dropin', {
  onReady: function (integration) {
    myIntegration = integration;
  }
});
	
// When you're ready to destroy your integration
myIntegration.teardown(function () {
  // braintree.setup can safely be run again!
});
```

This `teardown` method will go ahead and do the work necessary to destroy anything that braintree has created on your page including `iframes` and hidden `input` elements. It will also close any popups or modals that have been opened by our integration (PayPal). `teardown` will execute a callback when these operations have completed and you can safely proceed with your next steps.

**Comment by berendberendsen** [comment]
@kyledetella @rturumella just tested this and it seems to be working fine. I added a `$scope.$on('$destroy'` bit in the code to call the teardown() and it seems to have nicely cleaned up the dropin. Thanks for the resolution in the end!

**Comment by cityreader** [comment]
@rturumella Does this update also apply to hosted fields js?

**Comment by kyledetella** [comment]
@cityreader this update will apply to `hosted-fields` but that has not been released yet. We are aggressively working to provide `teardown` for `hosted-fields` integrations as well. That work lives in a separate beta build and there are a few things we need to reconcile first. I will keep you posted with updates on that.

**Comment by theotherzach** [comment]
@kyledetella How are we doing on the hosted fields integration?

**Comment by kyledetella** [comment]
@theotherzach the `teardown` functionality in `hosted-fields` will be available within 2 weeks. Again, we will make sure to drop the news here as soon as its ready.

Have you had a chance to try out `teardown` in your Drop-in integration? Is it working as expected for you?

**Comment by rbsmidt** [comment]
@kyledetella sounds good - looking forward to the new beta.
Just to clarify: I'm currently testing out hosted fields on a React JS powered check out flow. I'm currently having a hard time getting my braintree.setup to fetch my form correct. I do get the onReady-callback (but with no arguments passed - guessing it's because the integration-argument isn't included in the newest hosted-fields beta build?). But for instance my placeholders and other options aren't applied to the input fields in question. Also the onFieldEvent isn't firing when changing value in input fields. When i try to submit a transaction, my onError-event fires with this msg:

    message: "User did not enter a payment method"
    type: "VALIDATION"

Also my onFieldEvent fires for all inputs on submit - but just after the error msg. They all appear to be empty according to the onFieldEvent, and the container is ".braintree-hosted-fields-invalid" - which obvously tells me something is wrong with my braintree init.

I'm guessing the above is caused by the way react is rendering js and markup. Will i be able to control this better, when the teardown-function is added to braintree?
And is it correct understood, that with the current build of hosted fields, it is not possible to load the hosted fields dynamically on a onpage javascript application?

Thanks for keeping us all updated through development!

**Comment by kyledetella** [comment]
@rbsmidt  the `onReady` callback will not pass any arguments until `teardown` is introduced into Hosted Fields, so that is expected.

As for the other issues you mentioned (placeholders and options not being applied), these sound unrelated to `teardown`. It will be easier for us to debug and help you out if we could get some more context on your integration. Would you be willing to post snippets of your setup? If you would rather not post them here, send us an email to hosted-fields@getbraintree.com and we can take this conversation there.

Regarding your last question, let me make sure I understand, are you asking if it is possible to setup hosted fields dynamically after page load? If so, yes you are able to do that. If I am misunderstanding the question, let me know.

**Comment by rbsmidt** [comment]
Thanks for your reply @kyledetella 

Yes i've been debugging a bit more, and it seems that you are right, my issue lies elsewhere. I'm happy to provide an example:

```
var Checkout = React.createClass({
  componentDidMount: function(){
    console.log('checkout mounted..');
    this.initPayment();
  },
  initPayment: function(){
    console.log('init payment', this);
    var options = {
      id: "payment-form",
      hostedFields: {
        onFieldEvent: function(event){
          console.log('field event triggerede', event);
        },
        number: {
          selector: '#card-number',
          placeholder: 'Kortnummer'
        },
        cvv: {
          selector: '#cvv',
          placeholder: 'CVV'
        },
        expirationDate: {
          selector: '#expiration-date',
          placeholder: 'Kortnummer'
        }
      },
      onReady: function(){
        console.log('braintree ready');
      },
      onError: function(error){
        console.log('an error occurred', error);
      },
      onPaymentMethodReceived: function(event, data){
        console.log('event', event, 'data', data);
      }
    }
    braintree.setup(this.props.state.token, "custom", options);
  },
  render: function(){
    console.log('render checkout', this);
    return(
        <div className="grid-checkout">
          <h3>Check ud</h3>
          <form id="payment-form">
              <input type="text" name="firstname" id="fistname" placeholder="Fornavn" />
              <input type="text" name="lastname" id="lastname" placeholder="Efternavn" />
              <input type="text" name="email" id="email" placeholder="Email" />
              <input type="text" name="tlf" id="tlf" placeholder="Telefon" />
              <p><img src="/images/icon_security.svg" height="15" width="auto" /> Sikker betaling <img src="/images/cc@2x.png" width="auto" height="25"/></p>
              <input type="text" name="card-number" id="card-number"  />
              <input type="text" name="cvv" id="cvv" />
              <input type="text" name="expiration-date" id="expiration-date" />
              <button>Betal</button>
          </form>
        </div>
    );
  }
});
```

As stated in my last comment, i the onReady-event fires successfully after componentDidMount and the initPayment, but my placeholders doesn't show up, and the onFieldEvent doesn't fire either. When i press the submit-button, my three hosted fields gets the class `braintree-hosted-fields-invalid`. But before i press submit, the input-fields doesn't have a class. So it seems to me, that my braintree setup never "catches" the three input fields - until i try submitting, and somehow they are registered as empty fields by braintree.js. 
The error message is still as stated in my last comment.

Any idea what could cause this setup issue?

Best regards,
Rasmus

**Comment by lilaconlee** [comment]
@rbsmidt It looks like if you replace the 	`input` elements for `card-number`, `cvv`, and `expiration-date` with `div`s, everything works. Hosted Fields actually uses these `div` elements as containers in which we inject iframes which contain the "hosted fields". Take a look at the [documentation](https://developers.braintreepayments.com/javascript/guides/hosted-fields/setup-and-integration) and/or this [blog post](https://www.braintreepayments.com/blog/your-checkout-our-iframes/) for more detailed information on setting up Hosted Fields.

Here is an example of the updated code for you to try out:

```js
...
  render: function(){
    console.log('render checkout', this);
    return(
        <div className="grid-checkout">
          <h3>Check ud</h3>
          <form id="payment-form">
              <input type="text" name="firstname" id="fistname" placeholder="Fornavn" />
              <input type="text" name="lastname" id="lastname" placeholder="Efternavn" />
              <input type="text" name="email" id="email" placeholder="Email" />
              <input type="text" name="tlf" id="tlf" placeholder="Telefon" />
              <p><img src="/images/icon_security.svg" height="15" width="auto" /> Sikker betaling <img src="/images/cc@2x.png" width="auto" height="25"/></p>
              
              <div id="card-number"></div>
              <div id="cvv"></div>
              <div id="expiration-date"></div>
              
              <button>Betal</button>
          </form>
        </div>
    );
  }
});
...
```

I hope this helps you out. Let us know if you have any more questions.

**Comment by rbsmidt** [comment]
@lilaconlee Ahh you are totally right. I tested with divs instead of inputs, and it worked like a charm! Thanks a bunch.

**Comment by theotherzach** [comment]
@kyledetella we needed to change our drop-in solution out for hosted fields because of styling concerns.

**Comment by davidbryant** [comment]
Hi,

I've just been playing with the beta version and found that the onReady callback is not called in some cases, meaning that you aren't provided with the teardown function to clean it up. Specifically if the form is hidden at the time that setup() is called, onReady is not called until you that form becomes visible. If it never becomes visible then you never get the callback so cannot unregister the event handlers.

I've run into this because I have a multi-step check out process with the dropin UI on the second page, so I was loading it in the background since it can take a while to load. 

Is this intentional? If so why?

Cheers,
Dave

**Comment by kyledetella** [comment]
@davidbryant thanks for bringing this up. You are correct that `onReady` will not fire in the case of Drop-in until the UI has fully rendered (including the completion of css transitions). When the container/form for Drop-in is hidden, that action will be delayed until the surrounding DOM is made visible.

This is by design since the integration is technically not "ready" until it has fully rendered and is capable of tokenizing payment methods.

The concept of wanting to "preload" the Drop-in is interesting and I am glad you brought it up. The semantics of `onReady` suit the original intention of the Drop-in UI, but hearing about interesting use cases like yours really help as we make considerations for our products and features. I will bring this issue up with the team and let you know where we end up. 

To be honest though, it is unlikely we will change this pattern anytime soon, so in the case of your UI, I would recommend initializing Drop-in only after you have rendered the containing view and tearing it down prior to un-rendering that view.

**Comment by davidbryant** [comment]
So this would also imply that I need to prevent all navigation within a SPA from the time that I call braintree setup() until it fires its onReady event? Otherwise there's a chance that a navigation event in that timeframe will prevent onReady from being fired and so not provide me with the teardown callback?

**Comment by mrak** [comment]
@davidbryant that is correct.

It's not a convenient restriction, but in integrations such as dropin or hosted-fields there is a lot of cross-domain machinery being asynchronously set in motion; cutting it off halfway through with a navigation event could lead to dangling DOM nodes, event handlers, or an unusable integration.

**Comment by kyledetella** [comment]
This afternoon version `2.14.0` was released. As you will see in the [CHANGELOG](https://github.com/braintree/braintree-web/blob/master/CHANGELOG.md#2140), this release takes Hosted Fields out of beta and introduces the `teardown` API into those integrations as well.

This thread was originally opened to engage a dialogue on this matter and it certainly served that purpose. Should you experience any specific issues with `teardown` or Hosted Fields, please open a separate issue so that we can address them accordingly.

**Comment by cheyner** [comment]
@kyledetella how should we handle BraintreeData.setup() when using teardown()? 

Right now I'm calling
```javascript
braintree.setup(this.props.paymentToken, "custom", {
  onReady: function(integration) {
    Checkout.braintreeIntegration = integration;
  },
  id: "payment-method-form",
  hostedFields: {...});
BraintreeData.setup(this.props.merchantId, 'payment-method-form', 'production');
```

And then later using Checkout.braintreeIntegration.teardown() as needed.  It's working well, but will there be complications if I'm repeatedly calling BraintreeData.setup()?

**Comment by kyledetella** [comment]
@cheyner it doesn't sound as though you are experiencing issues when repeatedly calling `BraintreeData.setup()` so it should be fine. However, should you encounter any hiccups, feel free to open a new issue about this and we'll take a look.

**Comment by bezhermoso** [comment]
@kyledetella Thank you, the `teardown` APi is working great for us.

Just one bug I encountered: the `.bt-overlay` doesn't seem to be removed after `teardown`, and an new one is created whenever the PayPal login screen is displayed. This results to the transluscent black overlay to get darker and darker every teardown/initialization cycle. I remedied it by removing `.bt-overlay` within the `teardown` callback. Not sure if this is happening to others, too.

**Comment by kyledetella** [comment]
@bezhermoso glad to hear `teardown` is working out!

Would you mind opening a separate issue regarding `.bt-overlay` not being removed? It will be easier to track and handle specific issues in more focused threads. In the meantime, we will take a look.

**Comment by kyledetella** [comment]
The primary concern for most participants in this thread has been addressed in last week's [`2.14.0`](https://github.com/braintree/braintree-web/blob/master/CHANGELOG.md#2140) release. In the interest of keeping the discussion as specific as possible, we're going to close this issue.

Please open new issues for any further Single Page App concerns or questions. Thanks!

**Comment by trainerbill** [comment]
Thanks for the functionality.  When will it be documented?  Can you now call .setup() then braintree.teardown() and .setup again?

**Comment by kyledetella** [comment]
@trainerbill we will provide full documentation once this release is [hotlinked](https://developers.braintreepayments.com/javascript/guides/client-sdk#use-the-direct-link). That should happen sometime this week.

For now, the explanation and code provided [above](https://github.com/braintree/braintree-web/issues/29#issuecomment-137555915) remains accurate.

> Can you now call .setup() then braintree.teardown() and .setup again?

Yes, that is a core allowance with this new feature!

**Comment by parreirat** [comment]
Some months ago we badly needed this teardown() feature. We were changing billing just now and it's great that it's now available! Thank you!

**Comment by kyledetella** [comment]
@parreirat glad to hear it helped out!

Also, good call on posting the documentation in this thread, that slipped my mind!
